core/src/main/java/io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java

```diff
@@ -0,0 +1,148 @@
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.GeoByteFunction;
+import io.questdb.griffin.engine.functions.UnaryFunction;
+import io.questdb.griffin.engine.functions.constants.Constants;
+import io.questdb.std.IntList;
+import io.questdb.std.NumericException;
+import io.questdb.std.ObjList;
+
+import static io.questdb.cairo.ColumnType.GEO_HASH_MAX_BITS_LENGTH;
+
+public class CastStrToGeoHashFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "cast(Sg)";
+    }
+
+    @Override
+    public Function newInstance(int position,
+                                ObjList<Function> args,
+                                IntList argPositions,
+                                CairoConfiguration configuration,
+                                SqlExecutionContext sqlExecutionContext) throws SqlException {
+        Function value = args.getQuick(0);
+        int argPosition = argPositions.getQuick(0);
+        int geoType = args.getQuick(1).getType();
+        return newInstance(argPosition, geoType, value);
+    }
+
+    public Function newInstance(int argPosition, int geoType, Function value) throws SqlException {
+        if (value.isConstant()) {
+            try {
+                final int bits = ColumnType.getGeoHashBits(geoType);
+                assert bits > 0 && bits < GEO_HASH_MAX_BITS_LENGTH + 1;
+                return Constants.getGeoHashConstantWithType(
+                        getGeoHashImpl(value.getStr(null), argPosition, bits),
+                        geoType
+                );
+
+            } catch (NumericException e) {
+                // throw SqlException if string literal is invalid geohash
+                // runtime parsing errors will result in NULL geohash
+                throw SqlException.position(argPosition).put("invalid GEOHASH");
+            }
+        }
+        return new Func(geoType, value, argPosition);
+    }
+
+    // todo: consolidate
+    private static long getGeoHashImpl(CharSequence value, int position, int typeBits) throws SqlException, NumericException {
+        if (value == null || value.length() == 0) {
+            return GeoHashes.NULL;
+        }
+        int actualBits = value.length() * 5;
+        if (actualBits < typeBits) {
+            throw SqlException.position(position)
+                    .put("string is too short to cast to chosen GEOHASH precision [len=").put(value.length())
+                    .put(", precision=").put(typeBits).put(']');
+        }
+        // Don't parse full string, it can be over 12 chars and result in overflow
+        int parseChars = (typeBits - 1) / 5 + 1;
+        long lvalue = GeoHashes.fromString(value, 0, parseChars);
+        return lvalue >>> (parseChars * 5 - typeBits);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).